File: core/src/main/scala/diplomacy/variant/standard/rule/Rule.scala
package diplomacy.variant.standard.rule

import diplomacy.board.Power
import diplomacy.rule.{ Rule => BaseRule, InvalidOrderMessage }
import diplomacy.variant.standard.rule

final class Rule[Turn_ <: Turn, Power_ <: Power](nextTurn: Turn_ => Turn_)
  extends BaseRule[rule.State[Turn_], Power_, MilitaryBranch, UnitStatus, ProvinceStatus, Order.Order[Power_], Result]
    with BaseRule.TypeHelper {
  final type Turn = Turn_
  private[this] case class PhaseRule(
    resolver: (Turn => Turn, Board, Set[Order]) => Either[InvalidOrderMessage, ResolvedResult],
    generator: OrderGenerator[Turn, Power],
    validator: Validator[Turn, Power]
  )

  private[this] val rules = {
    val movementResolver = new MovementResolver {
      type Turn = Turn_
      type Power = Power_
    }
    val retreatResolver = new RetreatResolver {
      type Turn = Turn_
      type Power = Power_
    }
    val buildResolver = new BuildResolver {
      type Turn = Turn_
      type Power = Power_
    }

    Map(
      Movement -> PhaseRule(
        (x, y, z) => movementResolver(y, z collect { case o: Order.MovementOrder[Power] => o }),
        new MovementOrderGenerator[Turn, Power],
        new MovementValidator[Turn, Power]
      ),
      Retreat -> PhaseRule(
        (x, y, z) => retreatResolver(x)(y, z),
        new RetreatOrderGenerator[Turn, Power],
        new RetreatValidator[Turn, Power]
      ),
      Build -> PhaseRule(
        (x, y, z) => buildResolver(x)(y, z),
        new BuildOrderGenerator[Turn, Power],
        new BuildValidator[Turn, Power]
      )
    )
  }

  protected[this] def resolveProcedure(
    board: Board, orders: Set[Order]
  ): Either[InvalidOrderMessage, ResolvedResult] = {
    val PhaseRule(resolver, generator, validator) =
      this.rules(board.state.phase)

    for {
      result <- resolver(nextTurn, board, orders).right
      result2 <-
        (this.rules(result.board.state.phase).generator.ordersToSkipPhase(result.board) map {
          orders => this.resolve(result.board, orders)
        } getOrElse (Right(result))).right
    } yield {
      new ResolvedResult(result2.board, result.result ++ result2.result)
    }
  }

  protected[this] def unitsRequiringOrder(board: Board): Set[DiplomacyUnit] = {
    this.rules(board.state.phase).validator.unitsRequiringOrder(board)
  }

  protected[this] def errorMessageOfOrder(board: Board)(
    order: Order
  ): Option[InvalidOrderMessage] = {
    this.rules(board.state.phase).validator.errorMessageOfOrder(board)(order)
  }
  protected[this] def errorMessageOfOrders(board: Board)(
    orders: Set[Order]
  ): Option[InvalidOrderMessage] = {
    this.rules(board.state.phase).validator.errorMessageOfOrders(board)(orders)
  }

  protected[this] def defaultOrderOf(board: Board)(unit: DiplomacyUnit): Order = {
    this.rules(board.state.phase).generator.defaultOrder(board)(unit)
  }
}

object Rule {
  trait TypeHelper extends BaseRule.TypeHelper {
    type Turn <: diplomacy.variant.standard.rule.Turn

    type State = diplomacy.variant.standard.rule.State[Turn]
    type MilitaryBranch = diplomacy.variant.standard.rule.MilitaryBranch
    type UnitStatus = diplomacy.variant.standard.rule.UnitStatus
    type ProvinceStatus = diplomacy.variant.standard.rule.ProvinceStatus
    type Order = Order.Order[Power]
    type Result = diplomacy.variant.standard.rule.Result
  }
}

File: lib/variant/standard/rule/rule.js
const { Rule: BaseRule, ResolvedResult } = require('../../../rule/package')
const { Movement, Retreat, Build } = require('./phase')
const MovementResolver = require('./movement-resolver')
const MovementValidator = require('./movement-validator')
const MovementOrderGenerator = require('./movement-order-generator')
const RetreatResolver = require('./retreat-resolver')
const RetreatValidator = require('./retreat-validator')
const RetreatOrderGenerator = require('./retreat-order-generator')
const BuildResolver = require('./build-resolver')
const BuildValidator = require('./build-validator')
const BuildOrderGenerator = require('./build-order-generator')

class PhaseRule {
  constructor (resolver, validator, orderGenerator) {
    this.resolver = resolver
    this.validator = validator
    this.orderGenerator = orderGenerator
  }
}

module.exports = class Rule extends BaseRule {
  constructor () {
    super()
    this.phaseRules = new Map([
      [
        Movement,
        new PhaseRule(new MovementResolver(), new MovementValidator(), new MovementOrderGenerator())
      ],
      [
        Retreat,
        new PhaseRule(new RetreatResolver(), new RetreatValidator(), new RetreatOrderGenerator())
      ],
      [
        Build,
        new PhaseRule(new BuildResolver(), new BuildValidator(), new BuildOrderGenerator())
      ]
    ])
  }

  resolveProcedure (board, orders) {
    const { resolver, validator, orderGenerator: generator } =
      this.phaseRules.get(board.state.phase)
    const r1 = resolver.resolve(board, orders)
    if (r1.err) {
      return r1
    }

    const result = r1.result
    const orders2 =
      this.phaseRules.get(result.board.state.phase).orderGenerator.ordersToSkipPhase(result.board)

    if (!orders2) {
      return r1
    }

    const r2 = this.resolve(result.board, orders2)
    if (r2.err) {
      return r2
    }

    const result2 = r2.result

    const orderResults = new Set([...result.results])
    result2.results.forEach(r => orderResults.add(r))
    return {
      result:
        new ResolvedResult(result2.board, orderResults, result.isFinished || result2.isFinished)
    }
  }

  unitsRequiringOrder (board) {
    return this.phaseRules.get(board.state.phase).validator.unitsRequiringOrder(board)
  }

  errorMessageOfOrder (board, order) {
    return this.phaseRules.get(board.state.phase).validator.errorMessageOfOrder(board, order)
  }

  errorMessageOfOrders (board, orders) {
    return this.phaseRules.get(board.state.phase).validator.errorMessageOfOrders(board, orders)
  }

  defaultOrderOf (board, unit) {
    return this.phaseRules.get(board.state.phase).orderGenerator.defaultOrderOf(board, unit)
  }
}
